specification/dartLangSpec.tex

```diff
@@ -3411,89 +3414,54 @@ \subsubsection{Constant Constructors}
-
-The key issue is how one treats the formal parameters of a constructor.
-
-If a constant constructor is invoked from a constant object expression, the actual arguments will be required to be constant expressions.
-Therefore, if we were assured that constant constructors were always invoked from constant object expressions, we could assume that the formal parameters of a constructor were compile-time constants.
-
-However, constant constructors can also be invoked from ordinary instance creation expressions (\ref{new}), and so the above assumption is not generally valid.
-
-Nevertheless, the use of the formal parameters of a constant constructor within the constructor is of considerable utility.
-The concept of potentially constant expressions is introduced to facilitate limited use of such formal parameters.
-Specifically, we allow the usage of the formal parameters of a constant constructor for expressions that involve built-in operators, but not for constant objects, lists and maps.
-This allows for constructors such as:
-}
-
-\begin{dartCode}
-\CLASS{} C \{
-  \FINAL{} x; \FINAL{} y; \FINAL{} z;
-  \CONST{} C(p, q): x = q, y = p + 100, z = p + q;
-\}
-\end{dartCode}
-
-\commentary{
-The assignment to \code{x} is allowed under the assumption that \code{q} is constant (even though \code{q} is not, in general a compile-time constant).
-The assignment to \code{y} is similar, but raises additional questions.
-In this case, the superexpression of \code{p} is \code{p + 100}, and it requires that \code{p} be a numeric constant expression for the entire expression to be considered constant.
-The wording of the specification allows us to assume that \code{p} evaluates to an integer.
-A similar argument holds for \code{p} and \code{q} in the assignment to \code{z}.
-
-However, the following constructors are disallowed:
-}
-
-\begin{dartCode}
-\CLASS{} D \{
-  \FINAL{} w;
-  \CONST{} D.makeList(p): w = \CONST{} [p]; // \comment{compile-time error}
-  \CONST{} D.makeMap(p): w = \CONST{} \{"help": q\}; // \comment{compile-time error}
-  \CONST{} D.makeC(p): w = \CONST{} C(p, 12); // \comment{compile-time error}
-\}
-\end{dartCode}
+If a non-redirecting generative constant constructor is declared by a class $C$,
+it is a compile-time error
+for an instance variable declared in $C$
+to have an initializing expression that is not a constant expression.
 
-\commentary{
-The problem is not that the assignments to \code{w} are not potentially constant; they are.
-However, all these run afoul of the rules for constant lists (\ref{lists}), maps (\ref{maps}) and objects (\ref{const}), all of which independently require their subexpressions to be constant expressions.
+\commentary{%
+A superclass of $C$ cannot have such an initializing expression either,
+because it must necessarily declare
+a non-redirecting generative constant constructor.%
```

Comment:
Technically not true. it can have a non-constant field initializer and no non-redirecting generative constant constructor. It's an error, but that's stated elsewhere (later).
So, rephrase: 

  A superclass of $C$ cannot have such an initializing expression either
  if it has a non-redirecting generative constant constructor, 
  as required by the super-invocation of the non-redirection generative constant
  constructor of this class in the following paragraph.

In general, don't state anything about what a program can or can't do. It can do everything, some of it causes errors.

Reply:
Done.